compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1558,9 +1558,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             // FIXME: We are currently creating two branches here in order to maintain
             // consistency. But they should be merged as much as possible.
             let fru_tys = if self.tcx.features().type_changing_struct_update {
-                let base_ty = self.check_expr(base_expr);
                 match adt_ty.kind() {
                     ty::Adt(adt, substs) if adt.is_struct() => {
+                        // Make an ADT with fresh inference substitutions. This
+                        // will allow us to guide inference along so that, e.g.
+                        // ```
+                        // let x = MyStruct<'a, B, const C: usize> {
+                        //    f: 1,
+                        //    ..Default::default()
+                        // };
+                        // ```
+                        // will have the default base expression constrained to
+                        // `MyStruct<'_, _, _>`, as opposed to just `_`... This
+                        // will allow us to then do a subtyping relation on all
+                        // of the `remaining_fields` below, per the RFC.
+                        let fresh_substs = self.fresh_substs_for_item(base_expr.span, adt.did());
+                        let base_ty = self.check_expr_has_type_or_error(
+                            base_expr,
+                            self.tcx.mk_adt(*adt, fresh_substs),
+                            |_| {},
+                        );
+                        let base_ty = self.shallow_resolve(base_ty);
                         match base_ty.kind() {
                             ty::Adt(base_adt, base_subs) if adt == base_adt => {
```

Comment:
can you now `bug!` if that isn#t the case? or at least `delay_span_bug` if it's `ty::Error`? would safe us 8 columns of indentation ^^

Reply:
ironically, I don't think `check_expr_has_type_or_error` actually returns `ty::Error` when the expected type != the checked type of the input expr, hence us needing to call `report_mismatched_types` on line before:1605/after:1624. I can easily check, though.

I will see if I can simplify this, though. It can at least be changed to a let-chain :)